// ngx-color-picker/dist/lib/formats.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var Cmyk = (function () {
    function Cmyk(c, m, y, k) {
        this.c = c;
        this.m = m;
        this.y = y;
        this.k = k;
    }
    return Cmyk;
}());
export { Cmyk };
function Cmyk_tsickle_Closure_declarations() {
    /** @type {?} */
    Cmyk.prototype.c;
    /** @type {?} */
    Cmyk.prototype.m;
    /** @type {?} */
    Cmyk.prototype.y;
    /** @type {?} */
    Cmyk.prototype.k;
}
var Hsla = (function () {
    function Hsla(h, s, l, a) {
        this.h = h;
        this.s = s;
        this.l = l;
        this.a = a;
    }
    return Hsla;
}());
export { Hsla };
function Hsla_tsickle_Closure_declarations() {
    /** @type {?} */
    Hsla.prototype.h;
    /** @type {?} */
    Hsla.prototype.s;
    /** @type {?} */
    Hsla.prototype.l;
    /** @type {?} */
    Hsla.prototype.a;
}
var Hsva = (function () {
    function Hsva(h, s, v, a) {
        this.h = h;
        this.s = s;
        this.v = v;
        this.a = a;
    }
    return Hsva;
}());
export { Hsva };
function Hsva_tsickle_Closure_declarations() {
    /** @type {?} */
    Hsva.prototype.h;
    /** @type {?} */
    Hsva.prototype.s;
    /** @type {?} */
    Hsva.prototype.v;
    /** @type {?} */
    Hsva.prototype.a;
}
var Rgba = (function () {
    function Rgba(r, g, b, a) {
        this.r = r;
        this.g = g;
        this.b = b;
        this.a = a;
    }
    return Rgba;
}());
export { Rgba };
function Rgba_tsickle_Closure_declarations() {
    /** @type {?} */
    Rgba.prototype.r;
    /** @type {?} */
    Rgba.prototype.g;
    /** @type {?} */
    Rgba.prototype.b;
    /** @type {?} */
    Rgba.prototype.a;
}
